lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1944,6 +1905,62 @@ export class UserAgentApplication {
+   * @param {ClientInfo} clientInfo clientInfo received as part of the response comprising of fields uid and utid.
+   * @param {IdToken} idToken idToken received as part of the response.
+   * @ignore
+   * @private
+   * @hidden
+   */
+  private saveAccessToken(response: AuthResponse, authority: string, parameters: any, clientInfo: string): AuthResponse {
+    let scope: string;
+    let accessTokenResponse = response;
+    const clientObj: ClientInfo = new ClientInfo(clientInfo);
+    // if the response contains "scope"
+    if (parameters.hasOwnProperty("scope")) {
+      // read the scopes
+      scope = parameters["scope"];
+      const consentedScopes = scope.split(" ");
+      // retrieve all access tokens from the cache, remove the dup scores
+      const accessTokenCacheItems = this.cacheStorage.getAllAccessTokens(this.clientId, authority);
+      for (let i = 0; i < accessTokenCacheItems.length; i++) {
+        const accessTokenCacheItem = accessTokenCacheItems[i];
+        if (accessTokenCacheItem.key.userIdentifier === response.account.userIdentifier) {
+          const cachedScopes = accessTokenCacheItem.key.scopes.split(" ");
+          if (Utils.isIntersectingScopes(cachedScopes, consentedScopes)) {
+            this.cacheStorage.removeItem(JSON.stringify(accessTokenCacheItem.key));
+          }
+        }
+      }
+      // Generate and cache accessTokenKey and accessTokenValue
+      const expiresIn = Utils.expiresIn(parameters[Constants.expiresIn]).toString();
+      const accessTokenKey = new AccessTokenKey(authority, this.clientId, scope, clientObj.uid, clientObj.utid);
+      const accessTokenValue = new AccessTokenValue(parameters[Constants.accessToken], response.idToken.rawIdToken, expiresIn, clientInfo);
+      this.cacheStorage.setItem(JSON.stringify(accessTokenKey), JSON.stringify(accessTokenValue));
+      accessTokenResponse.accessToken  = parameters[Constants.accessToken];
+      accessTokenResponse.expiresIn = expiresIn;
+      accessTokenResponse.scopes = consentedScopes;
+    }
+    // if the response does not contain "scope" - scope is usually client_id and the token will be id_token
+    else {
+      scope = this.clientId;
+
+      // Generate and cache accessTokenKey and accessTokenValue
+      const accessTokenKey = new AccessTokenKey(authority, this.clientId, scope, clientObj.uid, clientObj.utid);
+      // TODO: since there is no access_token, this is also set to id_token?
+      const accessTokenValue = new AccessTokenValue(parameters[Constants.idToken], parameters[Constants.idToken], response.idToken.expiration, clientInfo);
+      this.cacheStorage.setItem(JSON.stringify(accessTokenKey), JSON.stringify(accessTokenValue));
+      accessTokenResponse.scopes = [scope];
+      accessTokenResponse.accessToken = parameters[Constants.idToken];
+      accessTokenResponse.expiresIn = response.idToken.expiration;
```

Comment:
Can we create a new object and not mutate what is passed in, this causes side effects to the caller of the fn